src/Icicle/Common/Exp/Eval/Small.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+
+module Icicle.Common.Exp.Eval.Small
+
+where
+
+import qualified Data.Map as M
+
+import Icicle.Internal.Pretty
+import Icicle.Common.Base
+import Icicle.Common.Type
+import Icicle.Common.Value
+import Icicle.Common.Exp.Exp
+import Icicle.Common.Exp.Eval (EvalPrim)
+import qualified Icicle.Common.Exp.Eval as Big
+
+import P
+
+import Data.Either.Combinators (mapLeft)
+
+
+data RuntimeError n p
+  = RuntimeErrorPrimBadArgs p [Value n p]
+  | RuntimeErrorUnboundVar  (Name n)
+  | RuntimeErrorPrimBad p
+  | RuntimeErrorPrimEval (Big.RuntimeError n p)
+  deriving (Show, Eq)
+
+--------------------------------------------------------------------------------
+
+-- | Small step evaluation. Call-by-value for simplicity.
+--
+eval :: Eq n
+     => EvalPrim n p -> Exp n p
+     -> Either (RuntimeError n p) (Exp n p)
+eval evalPrim xx = case xx of
+  -- * Variables must have been substituted away.
+  XVar n
+    -> Left (RuntimeErrorUnboundVar n)
+
+  -- * Values are fully evaluated.
+  XValue{}
+    -> return xx
+
+  -- * App where left is lambda and right is value: substitute.
+  XApp (XLam n t1 e) (XValue t2 v)
+    | t1 == t2
+    -> subst n v t1 e
+
+  -- * App where left is lambda: evals right.
+  XApp e1@XLam{} e2
+    -> fmap (XApp e1) (go e2)
```

Comment:
Argh, I made a mistake and asked you to implement small step evaluation, but it turns out what I really want is a partial evaluator.
Whether it's small step or big step doesn't really matter after all, and evaluation order doesn't matter, but it should go down and perform any simplifications like beta reduction and primitives applied to just XValues.
(We don't have recursion or fixpoints so we can afford to evaluate anything)